Public/Src/Tools/Execution.Analyzer/Analyzers.Partner/ContentPlacementAnalyzer.cs

```diff
@@ -0,0 +1,532 @@
+            m_outputWriter.WriteEndObject();
+            // done...
+        }
+
+        private void WriteJsonPropertyToStream<T>(string name, T value)
+        {
+            m_outputWriter.WritePropertyName(name);
+            m_outputWriter.WriteValue(value);
+        }
+
+        public bool IncludePip(Pip pip)
+        {
+            return pip.PipType != PipType.HashSourceFile;
+        }
+
+        public override void Prepare()
+        {
+            string outputFile = $"{Path.Combine(Input.CachedGraphDirectory, m_relativeOutputFile)}";
+            Console.WriteLine($"{GetType().FullName} starting. Output will be written to [{outputFile}]");
+            m_outputWriter = new JsonTextWriter(new StreamWriter(outputFile));
+            Contract.Requires(File.Exists(outputFile), $"Could not create [{outputFile}]...");
+        }
+
+        public override void Dispose()
+        {
+            if (m_outputWriter != null)
+            {
+                m_outputWriter.Close();
+            }
+            base.Dispose();
+        }
+
+        public override void FileArtifactContentDecided(FileArtifactContentDecidedEventData data)
+        {
+            string extension = Path.GetExtension(data.FileArtifact.Path.ToString(CachedGraph.Context.PathTable));
+            AddToDictionaryWithListAsValue(m_artifactsPerExtension, extension, data.FileArtifact);
+            CountInDictionary(m_countFilesPerExtension, extension);
+            m_fileContentMap[data.FileArtifact] = data.FileContentInfo;
+        }
+
+        /// <inheritdoc />
+        public override void PipExecutionPerformance(PipExecutionPerformanceEventData data)
+        {
+            AddToDictionaryWithListAsValue(m_scheduledPipExecutionData, data.PipId.Value, data);
+        }
+
+
+        private IList<ValType> GetOrReturnNull<KeyType, ValType>(Dictionary<KeyType, IList<ValType>> dictionary, KeyType key)
+        {
+            if (dictionary.ContainsKey(key))
+            {
+                return dictionary[key];
+            }
+            else
+            {
+                return null;
+            }
+        }
+
+        private void AddToDictionaryWithListAsValue<KeyType, ValType>(Dictionary<KeyType, IList<ValType>> dictionary, KeyType key, ValType value)
+        {
+            if (dictionary.ContainsKey(key))
+            {
+                dictionary[key].Add(value);
+            }
+            else
+            {
+                dictionary.Add(key, new List<ValType>(new ValType[] { value }));
+            }
+        }
+
+        private void CountInDictionary<KeyType>(Dictionary<KeyType, int> dictionary, KeyType key)
+        {
+            if (dictionary.ContainsKey(key))
+            {
+                dictionary[key] += 1;
+            }
+            else
+            {
+                dictionary.Add(key, 1);
+            }
+        }
+
+    }
+
+    public sealed class BXLPipData
+    {
+        public string SemiStableHash { get; set; } = null;

```

Comment:
>null [](start = 53, length = 4)

In C# you don't have to (and in this codebase we prefer not to) initialyze default values like null and 0. Would you mind removing these?

Reply:
Note same applies below for `BxlArtifact`

---
In reply to: [314069220](https://github.com/microsoft/BuildXL/pull/734#discussion_r314069220) [](ancestors = 314069220)